### Quality Checks
#### 1. Initialize
As we have already flattened the JSON files into their respective .csv files along with converting the UNIX timestamp to a more recognizable format, we will begin to analyze and perform checks on other sections of the dataset.

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline 

df_users = pd.read_csv('data/users.csv')
display(df_users.head())
display(df_users.tail())

_id.$oid  active    createdDate.$date      lastLogin.$date  \
0  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   
1  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   
2  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   
3  5ff1e1eacfcf6c399c274ae6    True  2021-01-03 15:25:30  2021-01-03 15:25:30   
4  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   

       role signUpSource state  
0  consumer        Email    WI  
1  consumer        Email    WI  
2  consumer        Email    WI  
3  consumer        Email    WI  
4  consumer        Email    WI

_id.$oid  active    createdDate.$date  \
490  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
491  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
492  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
493  54943462e4b07e684157a532    True  2014-12-19 14:21:22   
494  54943462e4b07e684157a532    True  2014-12-19 14:21:22   

         lastLogin.$date         role signUpSource state  
490  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
491  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
492  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
493  2021-03-05 16:52:23  fetch-staff          NaN   NaN  
494  2021-03-05 16:52:23  fetch-staff          NaN   NaN

#### 2. Rename columns or variables
Now, we check the data dictionary and the types assigned to the dataset. Along with this, we will rename some of these columns for better readability.

In [2]:
df_users = df_users.rename(columns=
                           {
                               '_id.$oid': 'id', 'createdDate.$date': 'createdDate', 'lastLogin.$date': 'lastLogin' 
                           })
display(df_users.dtypes)

id              object
active            bool
createdDate     object
lastLogin       object
role            object
signUpSource    object
state           object
dtype: object

#### 3. Check for duplicate rows or records
Some common issues with a dataset are that it might contain multiple duplicated rows which presents an incorrect idea about the dimensions and the size of the dataset.

If we find duplicated rows, we will drop these duplicates to simplify our dataset. Moreover, duplicate rows could indicate stale  or incomplete data.

In [3]:
duplicate_rows_df = df_users[df_users.duplicated()]

print("Actual shape v/s Duplicated shape")
df_users.shape, duplicate_rows_df.shape

Actual shape v/s Duplicated shape


((495, 7), (283, 7))

In [4]:
print("Before dropping duplicates")
display(df_users.count())
df_users = df_users.drop_duplicates()

print("After dropping duplicates")
display(df_users.count())

Before dropping duplicates


id              495
active          495
createdDate     495
lastLogin       433
role            495
signUpSource    447
state           439
dtype: int64

After dropping duplicates


id              212
active          212
createdDate     212
lastLogin       172
role            212
signUpSource    207
state           206
dtype: int64

#### 4. Check for NA values (NaN values)
Generally, we check for NaN and NULL values in our dataset and we replace them with better values in a process called Imputation. At times, these values can be simply replaced by a '0' and other times, it is best to populate an average value to it.

In some scenarios, if these values are not harmful, they are kept as is.

In this case, NA values were not dropped. <u>Further action could be taken based on response from the business stakeholders.</u>

In [5]:
print("Original count including NA values")
display(df_users.count())

temp = df_users.dropna()

print("Count after dropping NA values")
display(temp.count())

print("Total nulls")
df_users.isnull().sum()

Original count including NA values


id              212
active          212
createdDate     212
lastLogin       172
role            212
signUpSource    207
state           206
dtype: int64

Count after dropping NA values


id              165
active          165
createdDate     165
lastLogin       165
role            165
signUpSource    165
state           165
dtype: int64

Total nulls


id               0
active           0
createdDate      0
lastLogin       40
role             0
signUpSource     5
state            6
dtype: int64

#### 5. Individual checks for each column using unique(), min(), max()
We will now embark on analyzing each of these attributes/columns individually. Find out their range if data is numeric, if the values are categorical, we check to see if the categories mentioned are correctly stored.

In [6]:
print("Distinct values of:")
for col in df_users.columns:
    print(col+'=' + str(len(pd.unique(df_users[col]))))

print()
print("Categorical data")
lets_check_for = ['active', 'role', 'signUpSource', 'state']
for col in lets_check_for:
    print(col)
    print(pd.unique(df_users[col]))
    print("-----------------------")

Distinct values of:
id=212
active=2
createdDate=212
lastLogin=172
role=2
signUpSource=3
state=9

Categorical data
active
[ True False]
-----------------------
role
['consumer' 'fetch-staff']
-----------------------
signUpSource
['Email' 'Google' nan]
-----------------------
state
['WI' 'KY' 'AL' 'CO' 'IL' nan 'OH' 'SC' 'NH']
-----------------------


In [7]:
print("Range of createdDate values")
df_users["createdDate"].astype('datetime64[ns]').max(), df_users["createdDate"].astype('datetime64[ns]').min()

Range of createdDate values


(Timestamp('2021-02-12 14:11:06'), Timestamp('2014-12-19 14:21:22'))

In [8]:
print("Range of lastLogin values")
df_users["lastLogin"].astype('datetime64[ns]').max(), df_users["lastLogin"].astype('datetime64[ns]').min()

Range of lastLogin values


(Timestamp('2021-03-05 16:52:23'), Timestamp('2018-05-07 17:23:40'))

#### Conclusion
Criteria for measuring data quality:
##### 1. Accuracy
<u>a.</u> Duplication of records was noticed in the dataset. Total duplicate records in the dataset = 283. These records were dropped eventually. However, this could be an indicator of stale or inaccurate data.

<u>ACTION</u> - Reach out to the business stakeholders as well as the source system teams about the duplicates.

<u>b.</u> Almost all the other columns mentioned in the data dictionary met the required criteria except for the role column. The role column is a constant set to "consumer”; however, we can also see that some rows have "fetch-staff" value present.

<u>ACTION</u> - Ask the business stakeholders whether this is expected. Presumably, some of the actual FR staff have accessed this data and hence, this value is assigned to the records edited by the FR staff.

##### 2. Relevancy
<u>a.</u> Ask the business stakeholders whether this is expected. Presumably, some of the actual FR staff have accessed this data and hence, this value is assigned to the records edited by the FR staff.

##### 3. Completeness
<u>a.</u> Incomplete records were found as there were multiple columns with NA/NaN and NULL values present. This could be a problem from the source of the dataset itself.  

<u>ACTION</u> - Ask the source system team to investigate this issue with missing values or data.

##### 4. Timeliness
<u>a.</u> Based on the 2 columns that provide date-based information, the latest update in the dataset is from 2021-03-05.

<u>ACTION</u> - Need confirmation on whether this is the latest cut-off date for the dataset or if more recent data somehow went missing, pending further investigation.

##### 5. Consistency
<u>a.</u> No major issues with consistency of the data. Minor issue - date attributes are present in UNIX time. Need to confirm whether this is expected and consistent with other datasets. All the other files also were in the same UNIX time format which makes this a non-issue.

<u>ACTION</u> - Ask the business stakeholders and the source system team why the date format is set to UNIX time format.

##### References
Data Dictionary - https://fetch-hiring.s3.amazonaws.com/data-analyst/ineeddata-data-modeling/data-modeling.html